File: app/src/main/java/com/glookogluroo/bridge/nightscout/TreatmentMapper.kt
package com.glookogluroo.bridge.nightscout

import com.glookogluroo.bridge.glooko.BolusEntry
import com.glookogluroo.bridge.glooko.PumpStatistics
import java.time.Instant
import java.time.format.DateTimeFormatter
import kotlin.random.Random

object TreatmentMapper {
    private val isoFormatter = DateTimeFormatter.ISO_INSTANT

    fun bolusToTreatment(bolus: BolusEntry): NightscoutTreatment {
        return NightscoutTreatment(
            eventType = "Correction Bolus",
            insulin = bolus.units,
            carbs = bolus.carbsInput,
            enteredBy = NightscoutTreatment.ENTERED_BY,
            createdAt = isoFormatter.format(bolus.timestamp),
        )
    }

    fun bolusToContextNote(bolus: BolusEntry): NightscoutTreatment? {
        val notes = buildBolusContextNotes(bolus) ?: return null
        return NightscoutTreatment(
            eventType = "Note",
            enteredBy = NightscoutTreatment.ENTERED_BY,
            notes = notes,
            createdAt = isoFormatter.format(bolus.timestamp),
        )
    }

    fun bolusUploadTreatments(
        bolus: BolusEntry,
        jitterTimestamps: Boolean = false,
    ): List<NightscoutTreatment> {
        var bolusTreatment = bolusToTreatment(bolus)
        var contextNote = bolusToContextNote(bolus)
        if (jitterTimestamps) {
            bolusTreatment = applyInsulinTimestampJitter(bolusTreatment)
        }
        contextNote = contextNote?.copy(createdAt = oneSecondAfter(bolusTreatment.createdAt))
        return buildList {
            add(bolusTreatment)
            contextNote?.let { add(it) }
        }
    }

    fun isPumpModeNote(treatment: NightscoutTreatment): Boolean {
        return treatment.eventType == "Note" &&
            treatment.notes?.startsWith("Pump mode:") == true
    }

    fun applyInsulinTimestampJitter(
        treatment: NightscoutTreatment,
        jitterMicros: Long? = null,
    ): NightscoutTreatment {
        if (!treatment.eventType.endsWith("Bolus")) return treatment

        val micros = jitterMicros ?: randomJitterMicros()
        val instant = Instant.parse(treatment.createdAt)
        val jitteredAt = isoFormatter.format(instant.plusNanos(micros * 1_000))
        return treatment.copy(createdAt = jitteredAt)
    }

    private fun randomJitterMicros(): Long {
        var micros = Random.nextLong(-999_999, 1_000_000)
        if (micros == 0L) {
            micros = if (Random.nextBoolean()) 1 else -1
        }
        return micros
    }

    fun pumpModeToNote(pump: PumpStatistics, timestamp: Instant = Instant.now()): NightscoutTreatment {
        val modeLabel = pump.mode?.name?.lowercase() ?: "unknown"
        val notes = buildString {
            append("Pump mode: $modeLabel")
            append(" (auto ${pump.autoPercentage.toInt()}%")
            append(", manual ${pump.manualPercentage.toInt()}%")
            append(", limited ${pump.limitedPercentage.toInt()}%)")
            pump.totalInsulinPerDay?.let { append(" | ${"%.1f".format(it)} u/day") }
        }
        return NightscoutTreatment(
            eventType = "Note",
            enteredBy = NightscoutTreatment.ENTERED_BY,
            notes = notes,
            createdAt = isoFormatter.format(timestamp),
        )
    }

    fun deduplicationKey(treatment: NightscoutTreatment): String {
        val insulin = treatment.insulin ?: 0.0
        val carbs = treatment.carbs ?: 0.0
        return "${treatment.eventType}|${treatment.createdAt}|$insulin|$carbs|${treatment.notes.orEmpty()}"
    }

    private fun oneSecondAfter(createdAt: String): String {
        return isoFormatter.format(Instant.parse(createdAt).plusSeconds(1))
    }

    private fun buildBolusContextNotes(bolus: BolusEntry): String? {
        val parts = buildList {
            bolus.insulinOnBoard?.let { add("IOB: ${"%.2f".format(it)}u") }
            bolus.bloodGlucoseInput?.let { bg ->
                val source = bolus.bloodGlucoseSource?.takeIf { it.isNotBlank() } ?: "pump"
                add("CGM: $bg ($source)")
            }
        }
        return parts.joinToString(" | ").ifBlank { null }
    }
}

File: backend/g2g/treatment_mapper.py
"""Bolus → Nightscout treatment mapping — ported from TreatmentMapper.kt."""

from __future__ import annotations

import random
from datetime import datetime, timedelta, timezone

from g2g.models import BolusEntry, NightscoutTreatment, PumpStatistics


def _iso(dt: datetime) -> str:
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


def bolus_to_treatment(bolus: BolusEntry) -> NightscoutTreatment:
    return NightscoutTreatment(
        event_type="Correction Bolus",
        insulin=bolus.units,
        carbs=bolus.carbs_input,
        entered_by=NightscoutTreatment.ENTERED_BY,
        created_at=_iso(bolus.timestamp),
    )


def _build_bolus_context_notes(bolus: BolusEntry) -> str | None:
    parts: list[str] = []
    if bolus.insulin_on_board is not None:
        parts.append(f"IOB: {bolus.insulin_on_board:.2f}u")
    if bolus.blood_glucose_input is not None:
        source = bolus.blood_glucose_source.strip() if bolus.blood_glucose_source else "pump"
        if not source:
            source = "pump"
        parts.append(f"CGM: {bolus.blood_glucose_input} ({source})")
    return " | ".join(parts) if parts else None


def bolus_to_context_note(bolus: BolusEntry) -> NightscoutTreatment | None:
    notes = _build_bolus_context_notes(bolus)
    if not notes:
        return None
    return NightscoutTreatment(
        event_type="Note",
        entered_by=NightscoutTreatment.ENTERED_BY,
        notes=notes,
        created_at=_iso(bolus.timestamp),
    )


def apply_insulin_timestamp_jitter(
    treatment: NightscoutTreatment,
    jitter_micros: int | None = None,
) -> NightscoutTreatment:
    if not treatment.event_type.endswith("Bolus"):
        return treatment
    micros = jitter_micros if jitter_micros is not None else _random_jitter_micros()
    instant = datetime.fromisoformat(treatment.created_at.replace("Z", "+00:00"))
    jittered = instant + timedelta(microseconds=micros)
    return NightscoutTreatment(
        event_type=treatment.event_type,
        insulin=treatment.insulin,
        carbs=treatment.carbs,
        entered_by=treatment.entered_by,
        notes=treatment.notes,
        created_at=_iso(jittered),
    )


def _random_jitter_micros() -> int:
    micros = random.randint(-999_999, 999_999)
    if micros == 0:
        micros = 1 if random.choice([True, False]) else -1
    return micros


def _one_second_after(created_at: str) -> str:
    instant = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
    return _iso(instant + timedelta(seconds=1))


def bolus_upload_treatments(
    bolus: BolusEntry,
    jitter_timestamps: bool = False,
) -> list[NightscoutTreatment]:
    bolus_treatment = bolus_to_treatment(bolus)
    context_note = bolus_to_context_note(bolus)
    if jitter_timestamps:
        bolus_treatment = apply_insulin_timestamp_jitter(bolus_treatment)
    if context_note is not None:
        context_note = NightscoutTreatment(
            event_type=context_note.event_type,
            insulin=context_note.insulin,
            carbs=context_note.carbs,
            entered_by=context_note.entered_by,
            notes=context_note.notes,
            created_at=_one_second_after(bolus_treatment.created_at),
        )
    result = [bolus_treatment]
    if context_note is not None:
        result.append(context_note)
    return result


def is_pump_mode_note(treatment: NightscoutTreatment) -> bool:
    return treatment.event_type == "Note" and (treatment.notes or "").startswith("Pump mode:")


def pump_mode_to_note(
    pump: PumpStatistics,
    timestamp: datetime | None = None,
) -> NightscoutTreatment:
    ts = timestamp or datetime.now(timezone.utc)
    mode_label = pump.mode.value.lower() if pump.mode else "unknown"
    notes = (
        f"Pump mode: {mode_label}"
        f" (auto {int(pump.auto_percentage)}%"
        f", manual {int(pump.manual_percentage)}%"
        f", limited {int(pump.limited_percentage)}%)"
    )
    if pump.total_insulin_per_day is not None:
        notes += f" | {pump.total_insulin_per_day:.1f} u/day"
    return NightscoutTreatment(
        event_type="Note",
        entered_by=NightscoutTreatment.ENTERED_BY,
        notes=notes,
        created_at=_iso(ts),
    )


def deduplication_key(treatment: NightscoutTreatment) -> str:
    insulin = treatment.insulin if treatment.insulin is not None else 0.0
    carbs = treatment.carbs if treatment.carbs is not None else 0.0
    notes = treatment.notes or ""
    return f"{treatment.event_type}|{treatment.created_at}|{insulin}|{carbs}|{notes}"
